### PR TITLE
[Perf][MiniCPM-o] Avoid GPU sync in Resampler pos-cache adjust

### PR DESCRIPTION
**Pull request by amy-why-3459**
## Summary
- Stage0 torch profiles showed `Resampler._adjust_pos_cache` spending ~900ms+ on `cudaStreamSynchronize` / `aten::item`, because it compared a **GPU scalar** `max(tgt_sizes)` against host ints in a Python `if`.
- That sync landed **after** SigLIP, so vision-encoder GPU work was drained into the cache-adjust path and inflated `get_vision_hidden_states`.
- Pass **host `int` max height/width** into `_adjust_pos_cache`, materialize patch geometry on CPU (`tgt_sizes_cpu`), and **pre-grow** the 2D sincos pos cache **before** launching `vpm`, so cache growth cannot insert a mid-pipeline device sync.

## Changes
- `Resampler._adjust_pos_cache(max_h: int, max_w: int, device)` — no GPU tensor comparisons.
- `Resampler._tgt_size_lists` — host lists for `h` / `w` / `patch_len`.
- `get_vision_hidden_states` — host-side `tgt_sizes_cpu` + early `_adjust_pos_cache` / `pos_embed` device move before SigLIP; pass CPU `tgt_sizes` into `vpm` / `resampler`.

## Perf (stage0, MiniCPM-o 4.5)

| Metric | Before | After sync fix |
|--------|--------|----------------|
| `_adjust_pos_cache` | ~943 ms | ~0.02 ms |
| `cudaStreamSynchronize` (vision path) | ~950 ms | ~0.6 ms |
| Sync root cause | `if tensor > int` → `aten::item` | host ints only |

Remaining Resampler time is real compute (cross-attn / GEMM), not this sync bug.

## Test plan
- [ ] Restart MiniCPM-o 4.5 serve with torch profiler; confirm `_adjust_pos_cache` / `cudaStreamSynchronize` no longer dominate `get_vision_hidden_states`.
- [ ] Smoke image+text (and multimodal) request: embeddings and generation still succeed.
- [ ] Cover large `tgt_sizes` that force pos-cache growth (beyond default `max_size`) to ensure the host-int growth path works.

### CODE REVIEW

File: vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni_llm.py [hunk 1]
```diff
@@ -4120,9 +4131,17 @@ def get_vision_hidden_states(self, data: MiniCPMVImagePixelInputs) -> torch.Tens
...
+        # growth never inserts a mid-pipeline cudaStreamSynchronize.
+        tgt_sizes_cpu = tgt_sizes.detach().to(device="cpu")
+        hs = [int(h) for h in tgt_sizes_cpu[:, 0].tolist()]
+        ws = [int(w) for w in tgt_sizes_cpu[:, 1].tolist()]
+        num_patches = [h * w for h, w in zip(hs, ws)]
+        max_patches = max(num_patches) if num_patches else 0
+
+        self.resampler._adjust_pos_cache(max(hs) if hs else 0, max(ws) if ws else 0, device=device)
+        if self.resampler.pos_embed.device != device:
+            self.resampler.pos_embed = self.resampler.pos_embed.to(device)
```
Comment (yuanheng-zhao) [inline vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni_llm.py:4197]: There's another redundant calling of `_adjust_pos_cache` in `Resampler.forward method`? Better check if we shall update the forward method and remove any redundant processings

File: vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni_llm.py [hunk 2]
```diff
@@ -4120,9 +4131,17 @@ def get_vision_hidden_states(self, data: MiniCPMVImagePixelInputs) -> torch.Tens
...
 
-        num_patches = tgt_sizes.prod(-1)
-        max_patches = num_patches.max().item()
-        assert isinstance(max_patches, int)
+        # Host-side patch geometry before launching SigLIP, so resampler cache
+        # growth never inserts a mid-pipeline cudaStreamSynchronize.
+        tgt_sizes_cpu = tgt_sizes.detach().to(device="cpu")
+        hs = [int(h) for h in tgt_sizes_cpu[:, 0].tolist()]
+        ws = [int(w) for w in tgt_sizes_cpu[:, 1].tolist()]
+        num_patches = [h * w for h, w in zip(hs, ws)]
```
Comment (yuanheng-zhao) [inline vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni_llm.py]: Please check if these compute are consistent with your helper func `_tgt_size_lists`. If so, can replace with that helper to remove redundancy

File: vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni_llm.py [hunk 3]
```diff
@@ -1719,13 +1719,25 @@ def _set_2d_pos_cache(self, max_size, device="cpu"):
...
 
-    def _adjust_pos_cache(self, tgt_sizes, device):
-        max_h = torch.max(tgt_sizes[:, 0])
-        max_w = torch.max(tgt_sizes[:, 1])
+    def _adjust_pos_cache(self, max_h: int, max_w: int, device):
+        """Grow the 2D sincos cache from host ints.
+
+        Comparing GPU scalars in ``if tensor > int`` forces ``aten::item`` /
+        ``cudaStreamSynchronize`` and stalls behind the preceding SigLIP
+        encoder (~900ms+ in stage0 profiles). Pass Python ints instead.
```
Comment (yuanheng-zhao) [inline vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni_llm.py]: nit: We don't include "~900ms+ in stage0..." in func comment commonly

Comment (amy-why-3459) [reply vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni_llm.py]: > nit: We don't include "~900ms+ in stage0..." in func comment commonly

Good catches, thanks!

- `_tgt_size_lists`: done — `get_vision_hidden_states` now reuses `self.resampler._tgt_size_lists(tgt_sizes_cpu)` instead of recomputing `hs/ws/num_patches` inline.
- docstring: done — removed the profiler-specific "~900ms+" number.
- Redundant `_adjust_pos_cache` in `Resampler.forward`: I'd prefer to keep it. `Resampler.forward` is a standalone entry point (it can be called without going through `get_vision_hidden_states`), so it should still grow its own pos cache to stay self-contained. After this PR `_adjust_pos_cache` is a pure host-int comparison (no `aten::item` / stream sync) and is idempotent, so the second call on the vision path is effectively free — it just guards the standalone case. Let me know if you'd rather I hoist the geometry into a single call site instead.
